### PR TITLE
Support Iframes in wrench

### PR DESCRIPTION
**Pull request by jrmuizel**
These patches get us to a state where we can handle gecko recordings better.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/692)

### CODE REVIEW

File: wrench/src/scene.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::collections::HashMap;
+use webrender_traits::{AuxiliaryLists, BuiltDisplayList, PipelineId, Epoch, ColorF};
+use webrender_traits::DisplayItem;
+use webrender_traits::LayerSize;
```
Comment (kvark) [inline wrench/src/scene.rs]: nit: combine with the prev line `use`

File: wrench/src/scene.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+/// A representation of the layout within the display port for a given document or iframe.
+#[derive(Debug)]
+pub struct ScenePipeline {
+    pub pipeline_id: PipelineId,
```
Comment (kvark) [inline wrench/src/scene.rs]: The contents of this struct can be made private. Actually, where is it used at all? I see the `pipeline_map` getting populated, but it's only being iterated by the keys (`for pipeline_id in scene.pipeline_map.keys()`).

File: wrench/src/yaml_frame_writer.rs [hunk 3]
```diff
@@ -210,11 +211,28 @@ pub struct YamlFrameWriter {
...
+    frame_writer: YamlFrameWriter,
+    scene: Scene,
+}
+
+impl YamlFrameWriterReceiver {
+    pub fn new(path: &Path) -> YamlFrameWriterReceiver {
+        YamlFrameWriterReceiver {
+            frame_writer: YamlFrameWriter::new(path),
+            scene: Scene::new(),
+
```
Comment (vvuk) [inline wrench/src/yaml_frame_writer.rs]: unnecessary blank line

File: wrench/src/yaml_frame_writer.rs [hunk 4]
```diff
@@ -311,21 +311,48 @@ impl YamlFrameWriter {
...
         let mut root = new_table();
-        table_node(&mut root, "root", root_dl_table);
+        if let Some(root_pipeline_id) = scene.root_pipeline_id {
+
+            u32_vec_node(&mut root_dl_table, "id", &vec![root_pipeline_id.0, root_pipeline_id.1]);
+
+            let mut pipelines = vec![];
+            for pipeline_id in scene.pipeline_map.keys() {
+                // write out all pipelines other than the root one
+                if pipeline_id.0 == root_pipeline_id.0 && pipeline_id.1 == root_pipeline_id.1 {
```
Comment (vvuk) [inline wrench/src/yaml_frame_writer.rs]: there's no PartialEq defined for this, can't just pipeline_id == root_pipeline_id?

File: wrench/src/yaml_frame_writer.rs [hunk 5]
```diff
@@ -311,21 +311,48 @@ impl YamlFrameWriter {
         scene.finish_root_display_list(self.pipeline_id.unwrap(), dl, aux);
 
         let mut root = new_table();
-        table_node(&mut root, "root", root_dl_table);
+        if let Some(root_pipeline_id) = scene.root_pipeline_id {
+
```
Comment (vvuk) [inline wrench/src/yaml_frame_writer.rs]: unnecessary blank line

File: wrench/src/yaml_frame_writer.rs [hunk 6]
```diff
@@ -311,21 +311,48 @@ impl YamlFrameWriter {
...
+                let mut pipeline = new_table();
+                u32_vec_node(&mut pipeline, "id", &vec![pipeline_id.0, pipeline_id.1]);
+
+                let dl = scene.display_lists.get(&pipeline_id).unwrap();
+                let aux = scene.pipeline_auxiliary_lists.get(&pipeline_id).unwrap();
+                let mut iter = dl.iter();
+                self.write_dl(&mut pipeline, &mut iter, &aux);
+                pipelines.push(Yaml::Hash(pipeline));
+            }
+
```
Comment (vvuk) [inline wrench/src/yaml_frame_writer.rs]: blank line

**Pull request merged**